irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -734,88 +753,94 @@ namespace iroha {
+      if (ledger_state_) {
+        return expected::makeValue(ledger_state_.value());
+      } else {
+        // throw? because state in iroha is inconsistent with DB
+        return expected::makeError(
+            "This should never happen - a missing ledger state after a "
+            "successfull commit!");
       }
     }
 
-    boost::optional<std::unique_ptr<LedgerState>> StorageImpl::commitPrepared(
+    bool StorageImpl::preparedCommitEnabled() const {
+      return prepared_blocks_enabled_ and block_is_prepared_;
+    }
+
+    CommitResult StorageImpl::commitPrepared(
         std::shared_ptr<const shared_model::interface::Block> block) {
       if (not prepared_blocks_enabled_) {
-        log_->warn("prepared blocks are not enabled");
-        return boost::none;
+        return expected::makeError(
+            std::string{"prepared blocks are not enabled"});
       }
 
       if (not block_is_prepared_) {
-        log_->info("there are no prepared blocks");
-        return boost::none;
+        return expected::makeError("there are no prepared blocks");
       }
+
       log_->info("applying prepared block");
 
       try {
         std::shared_lock<std::shared_timed_mutex> lock(drop_mutex_);
         if (not connection_) {
-          log_->info(
+          std::string msg(
               "commitPrepared: connection to database is not initialised");
-          return boost::none;
+          return expected::makeError(std::move(msg));
         }
         soci::session sql(*connection_);
         sql << "COMMIT PREPARED '" + prepared_block_name_ + "';";
         PostgresBlockIndex block_index(
             sql, log_manager_->getChild("BlockIndex")->getLogger());
         block_index.index(*block);
         block_is_prepared_ = false;
-        return PostgresWsvQuery(sql,
-                                factory_,
-                                log_manager_->getChild("WsvQuery")->getLogger())
-                       .getPeers()
-                   | [this, &block, &sql](auto &&peers)
-                   -> boost::optional<std::unique_ptr<LedgerState>> {
-          if (this->storeBlock(block)) {
-            return boost::optional<std::unique_ptr<LedgerState>>(
-                std::make_unique<LedgerState>(
-                    std::move(peers), block->height(), block->hash()));
+
+        return storeBlock(block) | [this, &sql, &block]() -> CommitResult {
+          decltype(
+              std::declval<PostgresWsvQuery>().getPeers()) opt_ledger_peers;
+          {
+            auto peer_query = PostgresWsvQuery(
+                sql,
+                this->factory_,
+                this->log_manager_->getChild("WsvQuery")->getLogger());
+            if (not(opt_ledger_peers = peer_query.getPeers())) {
+              return expected::makeError(
+                  std::string{"Failed to get ledger peers! Will retry."});
+            }
           }
-          return boost::none;
+          assert(opt_ledger_peers);
```

Comment:
Assertion can be removed for the same reason as in mutable storage. Explicit variable with decltype-declval will not be required in this case. The additional scope above can also be removed.